### PR TITLE
upstream: deprecate legacy load balancer creation code path

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: upstream: upstream: deprecate legacy load balancer creation path
Additional Description:

Now, we have two different way to configure and create upstream LB. One is the new load balancer extension point, one is the legacy API in the `Cluster`. These two ways have different code paths.

This PR did following things:
1. Create config wrapper for every LB. Every LB will has a legacy API wrapper and a normal API wrapper. Both wrappers are inherit from `LoadBalancerConfig`.
2. Refactor the LB extension factory. Now the LB factory could also take the (wrapped) legacy API. And legacy API is used, the legacy constructor of LB will be used.
3. Convert all legacy configuration to the config wrapper and extension factory will be used for all LB creation. This could be reverted by runtime flag.

Risk Level: low.
Testing: n/a.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

**Comment by wbpcode** [comment]
/wait

Will handle the CI failture tomorrow.

**Review by alyssawilk — COMMENTED** [review]
ugh sorry for delay - reviewed yesterday and failed to send :-/
/wait

File: source/extensions/load_balancing_policies/least_request/config.cc
Comment (wbpcode) [reply source/extensions/load_balancing_policies/least_request/config.cc:13]: > if I were picking this up I'd just move one class at a time from legacy construction methods to using factory classes.
if (config == hard_codec_cluster) {
factory_name = "registered.type"
}

I think that's where we are going to. But I have a longer path to achieve this target. I will create a config type for every lb. Then I will convert both the legacy proto API and new proto API to the config type.
Finally, a runtime flag and simliar logic like your preudo-code will be added.

Of course, there are other possible paths. Maybe we can also convert the legacy proto API to the new proto API directly. 🤔 I can create another PR to try it.

Comment (wbpcode) [reply source/extensions/load_balancing_policies/least_request/config.cc:13]: > I think I'm confused at the move from specifically typed config to generic config with dynamic cast.

Ignore it, here because I'm lazy.

Comment (wbpcode) [reply source/extensions/load_balancing_policies/least_request/config.cc:13]: /wait

Maybe convert the legacy proto API to the new proto API is more simple and easy to review/understand. I will do a try first. Before that, pend this PR.

File: source/common/upstream/load_balancer_impl.h
Comment (soulxu) [inline source/common/upstream/load_balancer_impl.h:587]: Can we change the EdfLoadBalancerBase's parameter type to `const absl::optional<>&`, then we needn't this copy anymore.

Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.h:587]: Of course. But this PR is suspended temporarily. So, you can also delay this review. Thanks.

File: source/common/upstream/BUILD
Comment (alyssawilk) [inline source/common/upstream/BUILD:429]: We moved these out to extensions so we didn't need the hard link.  I'd prefer we avoid adding them back in.

File: source/common/upstream/upstream_impl.cc
Comment (alyssawilk) [inline source/common/upstream/upstream_impl.cc:1219]: so generally we runtime guard if there's going to be a functional change in behavior.  Is this expected to be a behavioral change, or guarded just because it's high risk?

File: source/common/upstream/upstream_impl.h
Comment (alyssawilk) [inline source/common/upstream/upstream_impl.h:947]: optional nit: consider
return (lb_policy_config_ == nullptr) ? {} : lb_policy_config_->lbMaglevConfig();
if the compiler lets us.

File: source/extensions/load_balancing_policies/least_request/config.cc
Comment (alyssawilk) [reply source/extensions/load_balancing_policies/least_request/config.cc:13]: so can we get rid of it with the reloadable flag?  If so I'd be more inclined to just skip that step.

**Review by alyssawilk — COMMENTED** [review]
Cool, it's helpful to see the full context, as well as the final state for all of this!  some follow-up thoughts
/wait

File: source/common/upstream/BUILD
Comment (wbpcode) [reply source/common/upstream/BUILD:429]: got it, I can split it out

File: source/common/upstream/upstream_impl.cc
Comment (wbpcode) [reply source/common/upstream/upstream_impl.cc:1219]: Because the high risk. Ideally, this shouldn't bring any behaviour change.

File: source/extensions/load_balancing_policies/least_request/config.cc
Comment (wbpcode) [reply source/extensions/load_balancing_policies/least_request/config.cc:13]: Yeah, it is completely possible and I think we will finally get there.
But it is a little hard to make it as target of this patch.

We can parse both legacy API and new API to same typed config, and then, all these if check or dynamic cast could be removed. (Just like subset lb did.)

But this way need lots of change to the constructor which will make the patch bigger and hard to review.

So I prefer to do that after the old lb
code path is deprecated/removed.

Of course, if you think it's okay, I think i can do the constructor/config unification first.

Comment (alyssawilk) [reply source/extensions/load_balancing_policies/least_request/config.cc:13]: Would the risk here be a start-up crash, or load balancing performing differently?  If it's the former I'd be inclined to do it without flag reload.  If it's the latter then yes, caution makes more sense.

Comment (wbpcode) [reply source/extensions/load_balancing_policies/least_request/config.cc:13]: I think the risk would be ` load balancing performing differently` or `unexpected bug`. I prefer to use a runtime flag here then we will have six mouths to ensure the stability of new code path.

**Comment by wbpcode** [comment]
not ready for formal review.

/wait

File: source/common/upstream/upstream_impl.h
Comment (wbpcode) [reply source/common/upstream/upstream_impl.h:917]: no. compiler forbidden it.

**Comment by wbpcode** [comment]
Re: so can we get rid of it with the reloadable flag? If so I'd be more inclined to just skip that step.

cc @alyssawilk 

I think I mis-understood your mean before when I saw this message on my phone.

In fact, I am also very hesitant right now. In the one hand, I have no 100% confidence to the new code path. In the another hand, if reloadable flag is used, we can only do further refactor (move remaining lb code to the extensions/) to the LB after six mouths.

**Comment by wbpcode** [comment]
Waiting for https://github.com/envoyproxy/envoy/pull/29613

**Comment by wbpcode** [comment]
Hi, @alyssawilk , I think this PR is ready for a formal review. It super bigger than my initial expectation, mainly because the we need update lots of different BUILD files.

There are only three core logic updates:
1. Legacy API conversation: wrapper the legacy API to a typed wrapper. See `getTypedLbConfigFromLegacyProto`.
2. Update all LB extension factories to accept the legacy API wrapper. See related changes in the `/source/extensions/load_balancing_polices`.
3. Enable the API conversation by default and could be controlled by a runtime flag.

All other changes are because this PR introduces some deps changes. For example, in the previous implementation, the one who use round robin only depends on `upstream_lib`, now, it depends on `upstream_lib` and `/source/extensions/load_balancing_polices/round_robin`.

**Comment by alyssawilk** [comment]
yikes, yeah this is a doozy :-P  I'm still going to let @soulxu do first pass but I'll try to do an early pass next week whether he's done or not.

**Comment by alyssawilk** [comment]
ALso looks like the mobile compile time options fail is real.  I think there's afile to include protos?  @abeyad may be able to suggest debug tips since @RyanTheOptimist is out.

**Comment by abeyad** [comment]
> ALso looks like the mobile compile time options fail is real. I think there's afile to include protos? @abeyad may be able to suggest debug tips since @RyanTheOptimist is out.

The error is:
```
assert failure: reflectable_message. Details: Unable to create dyanmic message for: envoy.extensions.load_balancing_policies.cluster_provided.v3.ClusterProvided
```

Which is due to the lite proto code not having the ClusterProvided proto descriptor loaded (https://github.com/envoyproxy/envoy/blob/3890fbc25537d5c77dccd8f0cfac6e4db03f52f2/source/common/protobuf/create_reflectable_message.cc#L193).  I'm not sure why it's trying to access that proto though..

File: source/common/upstream/load_balancer_impl.h
Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.h]: Will defer this optimization to new PR or land it after the legacy code path is removed.
This PR is too big now, I don't want to add new complexity. orz.

**Comment by wbpcode** [comment]
> ALso looks like the mobile compile time options fail is real. I think there's afile to include protos? @abeyad may be able to suggest debug tips since @RyanTheOptimist is out.

> Which is due to the lite proto code not having the ClusterProvided proto descriptor loaded (

My negligence. I forgot to load after including it in the source file.

File: source/common/upstream/upstream_impl.cc
Comment (soulxu) [inline source/common/upstream/upstream_impl.cc:1169]: Why we always override the lb_type to `LoadBalancerType::LoadBalancingPolicyConfig`?

File: source/common/upstream/upstream_impl.cc
Comment (soulxu) [inline source/common/upstream/upstream_impl.cc:1181]: nit, it seems ok to not cleanup this, then you needn't all the changes in the upstream_impl.h

File: source/extensions/load_balancing_policies/maglev/config.cc
Comment (soulxu) [inline source/extensions/load_balancing_policies/maglev/config.cc:21]: Is it better to move those `dynamic_cast` to a method in the lb_config?

File: source/extensions/load_balancing_policies/maglev/config.cc
Comment (soulxu) [inline source/extensions/load_balancing_policies/maglev/config.cc:25]: You miss an assertion  ```ASSERT(typed_lb_confg != nullptr || legacy_lb_config != nullptr,
         "Invalid load balancing policy configuration for least request load balancer");``` which added to other extensions

File: source/common/upstream/upstream_impl.cc
Comment (soulxu) [inline source/common/upstream/upstream_impl.cc:1163]: The line 1211 won't reach anymore.

File: changelogs/current.yaml
Comment (wbpcode) [reply changelogs/current.yaml:97]: When we trigger an unexpected LB bugs or behavior change. This flag allow users to work around the LB change temporarily.
Because we have no 100% confidence to the new code path and the LB is very very very important.

File: source/common/upstream/upstream_impl.cc
Comment (wbpcode) [reply source/common/upstream/upstream_impl.cc:1169]: This will tell the cluster manager to create LB from the typed load balancer factory.

File: source/common/upstream/upstream_impl.cc
Comment (wbpcode) [reply source/common/upstream/upstream_impl.cc:1181]: This is for reducing unnecessary memory overhead. We have lots of related PR to reduce memory overhead of clusters. I don't want to break their achievements.
Although I am open to this problem because all these code will be removed after six months.

File: source/extensions/load_balancing_policies/maglev/config.cc
Comment (wbpcode) [reply source/extensions/load_balancing_policies/maglev/config.cc:21]: I don't think so. Note these dynamic cast is only necessary for the built-in LBs, like the round robin, random.
But the `LoadBalancerConfig` interface is not only for the built-in LBs.

File: source/extensions/load_balancing_policies/maglev/config.cc
Comment (wbpcode) [reply source/extensions/load_balancing_policies/maglev/config.cc:25]: Although it's weird, but maglev and ring hash needn't that. 🤣 Because some historical reasons, these factories (maglev and ring hash) are also be used in legacy code path and `nullptr` will be used as `lb_config` argument.

We can only add that back after six month, when the legacy code path is removed completely.

File: source/common/upstream/upstream_impl.cc
Comment (wbpcode) [reply source/common/upstream/upstream_impl.cc:1163]: got it. Thanks.

File: mobile/envoy_build_config/extension_registry.cc
Comment (alyssawilk) [inline mobile/envoy_build_config/extension_registry.cc:191]: yeah I don't think we need any of these - would prefer to keep them out unless it breaks tests

File: mobile/envoy_build_config/extensions_build_config.bzl
Comment (alyssawilk) [inline mobile/envoy_build_config/extensions_build_config.bzl:34]: ditto here

File: source/common/upstream/load_balancer_config.h
Comment (alyssawilk) [inline source/common/upstream/load_balancer_config.h:124]: are these just pure moves?
Also can we get better comments?

File: source/common/upstream/load_balancer_config.h
Comment (alyssawilk) [inline source/common/upstream/load_balancer_config.h:219]: What is this?

**Review by alyssawilk — COMMENTED** [review]
LG overall.  Still waiting on @soulxu to LGTM his pass and needs a main merge

File: mobile/envoy_build_config/extension_registry.cc
Comment (wbpcode) [reply mobile/envoy_build_config/extension_registry.cc:191]: got it.

File: mobile/envoy_build_config/extensions_build_config.bzl
Comment (wbpcode) [reply mobile/envoy_build_config/extensions_build_config.bzl:34]: got it. But, at least the round robin (default lb) and cluster_provided (for forward proxy) are necessary.

File: source/common/upstream/load_balancer_config.h
Comment (wbpcode) [reply source/common/upstream/load_balancer_config.h:124]: Yeah, this is pure move. Will try to update the comment.

File: source/common/upstream/load_balancer_config.h
Comment (wbpcode) [reply source/common/upstream/load_balancer_config.h:219]: It's another implementation of `LoadBalancerSubsetInfo` that moved from `subset_lb.cc/h`. It's implemented for some historical reason.

Let's merge it to `LoadBalancerSubsetInfoImpl`.

File: source/extensions/load_balancing_policies/random/BUILD
Comment (wbpcode) [reply source/extensions/load_balancing_policies/random/BUILD:15]: Got it.

**Review by soulxu — DISMISSED** [review]
Sorry for the review late, since in the travel of KubeCon CN this week. LGTM, thanks!

File: source/common/upstream/load_balancer_config.h
Comment (alyssawilk) [inline source/common/upstream/load_balancer_config.h:25]: so I'm (tentatively) ok with this for this PR, but I'm inclined to think that we should move all ring hash code (legacy and otherwise) to the ring hash extension, and same for all the others.
I think we'd want each factory having a "convert config" function and if we had legacy ring hash config, the function would create new style config, and for new style configs the convert function would be a no-op.  WdYT?

File: source/common/upstream/load_balancer_config.h
Comment (alyssawilk) [inline source/common/upstream/load_balancer_config.h:160]: ditto here - seems like someone who doesn't include the subset LB would ideally not compile this at all.

Comment (wbpcode) [reply source/common/upstream/load_balancer_config.h:160]: We can do that after the old path is removed completely.

Six months of waiting. Although I don't  want the waiting actually, but I also have no enough confidence to switch to new path directly.

Comment (wbpcode) [reply source/common/upstream/load_balancer_config.h:160]: 🤔 may be I should be more brave. This PR actually didn't change any lb logic. It just change the way to create lb instance...

I am hesitant now.

File: source/common/upstream/load_balancer_config.h
Comment (alyssawilk) [reply source/common/upstream/load_balancer_config.h:25]: I believe you're correct but can you explain why not?  Even if we can only move the others out it's still worth doing IMO
Sorry I do owe you a full pass but I'm on reduced hours this week - I'll try to do a full pass tomorrow and thanks for your patience.  Feel free to resassign if my turnaround is too slow :-(

Comment (wbpcode) [reply source/common/upstream/load_balancer_config.h]: > I believe you're correct but can you explain why not? Even if we can only move the others out it's still worth doing IMO

The `LoadBalancerSubsetInfoImpl` is referenced by the core old directly for now in the legacy code path. So, we cannot move it to the `extensions/` for now.
But I do agree you point, so I have moved others. :)

Comment (wbpcode) [reply source/common/upstream/load_balancer_config.h]: > Sorry I do owe you a full pass but I'm on reduced hours this week - I'll try to do a full pass tomorrow and thanks for your patience. Feel free to resassign if my turnaround is too slow :-(

Just take it easy. I am not in a hurry. And rather than to land this patch quickly, I prefer to make it better. I am willing to give it more time.

File: source/common/upstream/upstream_impl.cc
Comment (alyssawilk) [inline source/common/upstream/upstream_impl.cc:932]: I don't think we can assert this is non-null - if the factory is not compiled in this will crash immediately below, no?  I think we should gracefully handle failure (either return null/null or make the first a StatusOr

File: source/common/upstream/upstream_impl.cc
Comment (alyssawilk) [inline source/common/upstream/upstream_impl.cc:945]: ditto here.

File: source/common/upstream/upstream_impl.cc
Comment (alyssawilk) [inline source/common/upstream/upstream_impl.cc:1141]: comment why?

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc
Comment (alyssawilk) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.cc:44]: if a requested factory is compiled out shouldn't we gracefully fail (throw/status fail) rather than crash?

File: source/extensions/load_balancing_policies/maglev/maglev_lb.cc
Comment (alyssawilk) [inline source/extensions/load_balancing_policies/maglev/maglev_lb.cc:5]: why the runtime include?

File: source/extensions/load_balancing_policies/random/BUILD
Comment (alyssawilk) [inline source/extensions/load_balancing_policies/random/BUILD:17]: optional but I'd super appreciate a follow-up moving some of the tests for the infrequently used LB to the individual directories.  Obviously not worth it for round robin but might be doable for lesser used ones.

File: source/extensions/load_balancing_policies/random/config.h
Comment (alyssawilk) [inline source/extensions/load_balancing_policies/random/config.h:18]: class comment?

**Review by alyssawilk — COMMENTED** [review]
Looks good overall - just a few final thoughts
/wait

File: source/common/upstream/upstream_impl.cc
Comment (wbpcode) [reply source/common/upstream/upstream_impl.cc:932]: sgtm.

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc
Comment (wbpcode) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc:44]: Reasonable suggestion. But if there is invalid lb factory that is compiled out, the cluster loading will failed when we create the cluster info.
So, this ASSERT should never happen and this why I used RELEASE_ASSERT here. If it's triggered, than it a bug rather than compile issue.

File: source/extensions/load_balancing_policies/maglev/maglev_lb.cc
Comment (wbpcode) [reply source/extensions/load_balancing_policies/maglev/maglev_lb.cc:5]: We removed the directly dependency to the `upstream_lib` and `upstream_impl.h` in the `maglev_lb.h`. That change result in we must include the `runtime_features.h` manually.
I think this change is reasonable and make the dependency graph simlper.

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc
Comment (alyssawilk) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc:44]: My concern is if the assertion ever fails (the utility is refactored) we'll have a production crash where we're trying really hard to avoid having release assert. At least let's add a comment what should happen here.  best case it'd be nice to regression test having a redis e2e test which tries to create a non-linked cluster to future-proof against harmful refactors.

Comment (wbpcode) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc:44]: got it. I can replace the RELEASE_ASSERT with a `ENVOY_BUG` and add additional check in the constructor of `ClusterInfoImpl`.

But it's hard to add a test case to test it. As I said in previous comments, the constructor of `ClusterInfoImpl` will throw if there is no valid factory and the creation of cluster will fail. The `isClusterProvidedLb()` will never be called for a fail-created cluster.

Comment (wbpcode) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: In fact, we do similiar `RELEASE_ASSERT` in other positions. See

https://github.com/envoyproxy/envoy/blob/b31401c94476bc52743f25021c50de8490abe63b/source/common/upstream/cluster_manager_impl.cc#L948-L958

**Comment by wbpcode** [comment]
Any new concern? cc @alyssawilk :)

**Review by alyssawilk — APPROVED** [review]

**Comment by wbpcode** [comment]
I am always scared of merging such big PR. Only the runtime flag give me a little of sense of security.

**Comment by kyessenov** [comment]
I think we need to document better how to migrate from CLUSTER_PROVIDED xDS to the new version since I'm struggling myself, and probably others will too after this PR. Can we revise the docs to show what it's supposed to look like? ORIGINAL_DST snippets no longer seem to work.

**Comment by kyessenov** [comment]
Ah, I forgot to include `cluster_provided` extension in the Istio build.

**Comment by wbpcode** [comment]
Ideally, this PR won't change any exist behavior except some extensions are stripped out. 

Users could continue to use their familiar API. (We have no plan to deprecate the legacy API because that may result in incredible noise to our users.)

**Comment by kyessenov** [comment]
Agree, I misunderstood the impact. After I added the cluster_provided extension, everything works as usual.

**Pull request merged**
